ngraph_bridge/enable_variable_ops/ngraph_variable_update_ng_tensor_op.cc

```diff
@@ -0,0 +1,117 @@
+
+namespace ngraph_bridge {
+
+//---------------------------------------------------------------------------
+//  NGraphVariableUpdateNGTensorOp::ctor
+//---------------------------------------------------------------------------
+NGraphVariableUpdateNGTensorOp::NGraphVariableUpdateNGTensorOp(
+    OpKernelConstruction* context)
+    : OpKernel(context) {
+  OP_REQUIRES_OK(context, context->GetAttr("ngraph_graph_id", &ng_graph_id_));
+  OP_REQUIRES_OK(context, context->GetAttr("ngraph_variable_shared_name",
+                                           &ng_variable_shared_name_));
+
+  NGRAPH_VLOG(4) << "NGraphVariableUpdateNGTensorOp:: Constructor called for: "
+                 << def().name() << " ,Graph ID " << ng_graph_id_
+                 << " ngraph_variable_shared_name " << ng_variable_shared_name_;
+
+  OP_REQUIRES(context, IsRefType(context->input_type(0)),
+              errors::InvalidArgument("input[0] needs to be a ref type"));
+}
+
+//---------------------------------------------------------------------------
+//  ~NGraphVariableUpdateNGTensorOp
+//---------------------------------------------------------------------------
+NGraphVariableUpdateNGTensorOp::~NGraphVariableUpdateNGTensorOp() {
+  NGRAPH_VLOG(4) << "~NGraphVariableUpdateNGTensorOp::" << name() << endl;
+}
+
+//---------------------------------------------------------------------------
+// OpKernel::Compute
+//---------------------------------------------------------------------------
+void NGraphVariableUpdateNGTensorOp::Compute(OpKernelContext* context) {
+  std::ostringstream oss;
+  // Start event tracing
+  ngraph::Event event_compute(oss.str(), name(), "");
+  bool log_copies = false;
+  OP_REQUIRES_OK(context,
+                 IsNgraphTFLogTensorCopiesEnabled(ng_graph_id_, log_copies));
+  std::stringstream copy_log_str;
+  NGRAPH_VLOG(4) << "KERNEL[" << type_string() << "]: " << name() << "\n";
+  int number_of_copies = 0;
+  NGRAPH_VLOG(4) << "NGraphVariableUpdateNGTensorOp:: Compute called for: "
+                 << def().name() << " ,Graph ID " << ng_graph_id_ << "\n";
+
+  // Since we have ngraph_variable_shared_name as an attribute,
+  // we can use that to get the variable from the context
+  NGraphVar* var;
+
+  OP_REQUIRES_OK(context, context->resource_manager()->Lookup<NGraphVar>(
+                              context->resource_manager()->default_container(),
+                              ng_variable_shared_name_, &var));
+
+  // Set the output Ref Tensor at output_index to be an alias of the
+  // input Ref Tensor at input_index.
+  // REQUIRES: IsRefType(input_dtype(input_index)).
+  // REQUIRES: IsRefType(output_dtype(output_index)).
+  context->forward_ref_input_to_ref_output(0, 0);
+
+  if (var->copy_tf_to_ng()) {
+    number_of_copies++;
+    copy_log_str << " COPY_TF_TO_NG ";
+  }
+
+  copy_log_str << " Number of copies " << number_of_copies << "\n";
+  if (log_copies) {
+    cout << copy_log_str.str();
+  }
+
+  // Unref Var
+  var->Unref();
```

Comment:
We can add a comment here perhaps, that `Lookup` ultimately calls `DoLookup`, which calls `Ref` here: https://github.com/tensorflow/tensorflow/blob/5448f25041ed5d32b8aee08250e8ec66e7353593/tensorflow/core/framework/resource_mgr.cc#L187

hence we are `Unref`ing